src/Compilers/Core/MSBuildTask/CanonicalError.cs

```diff
@@ -0,0 +1,413 @@
+            parsedMessage.subcategory = match.Groups["SUBCATEGORY"].Value.Trim();
+
+            // Next, see if category is something that is recognized.
+            if (0 == String.Compare(category, "error", StringComparison.OrdinalIgnoreCase))
+            {
+                parsedMessage.category = Parts.Category.Error;
+            }
+            else if (0 == String.Compare(category, "warning", StringComparison.OrdinalIgnoreCase))
+            {
+                parsedMessage.category = Parts.Category.Warning;
+            }
+            else
+            {
+                // Not an error\warning message.
+                return null;
+            }
+
+            // Origin is not a simple file, but it still could be of the form,
+            //  foo.cpp(location)
+            match = filenameLocationFromOrigin.Match(origin);
+
+            if (match.Success)
+            {
+                // The origin is in the form,
+                //  foo.cpp(location)
+                // Assume the filename exists, but don't verify it. What else could it be?
+                string location = match.Groups["LOCATION"].Value.Trim();
+                parsedMessage.origin = match.Groups["FILENAME"].Value.Trim();
+
+                // Now, take apart the location. It can be one of these:
+                //      loc:
+                //      (line)
+                //      (line-line)
+                //      (line,col)
+                //      (line,col-col)
+                //      (line,col,len)
+                //      (line,col,line,col)
+                if (location.Length > 0)
+                {
+                    match = lineFromLocation.Match(location);
+                    if (match.Success)
+                    {
+                        parsedMessage.line = ConvertToIntWithDefault(match.Groups["LINE"].Value.Trim());
+                    }
+                    else
+                    {
+                        match = lineLineFromLocation.Match(location);
+                        if (match.Success)
+                        {
+                            parsedMessage.line = ConvertToIntWithDefault(match.Groups["LINE"].Value.Trim());
+                            parsedMessage.endLine = ConvertToIntWithDefault(match.Groups["ENDLINE"].Value.Trim());
+                        }
+                        else
+                        {
+                            match = lineColFromLocation.Match(location);
+                            if (match.Success)
+                            {
+                                parsedMessage.line = ConvertToIntWithDefault(match.Groups["LINE"].Value.Trim());
+                                parsedMessage.column = ConvertToIntWithDefault(match.Groups["COLUMN"].Value.Trim());
+                            }
+                            else
+                            {
+                                match = lineColColFromLocation.Match(location);
```

Comment:
Currently I'm simply moving the code "as is," but I will certainly keep that in mind when we start refactoring this code.